### ISSUE TITLE
Understanding distributed training implementation

### ISSUE THREAD

**Issue by pluiez** [pytorch/fairseq#582]
Hi,

Suppose that in the case of a single gpu, the batch size is at most B, and the logits returned by the model forward is of size (BxTxV). 

So when I use N GPUs, I scale the batch size to N*B. At the end of the model forward, the results of the N GPUs will be merged into the rank-0 GPU , so the rank-0 GPU needs to save the result of size (N*BxTxV). Doesn't this lead to out-of-memory error on rank-0 GPU? Why not calculate the loss and gradient separately for each GPU, then all_reduce the gradients?

How do you deal with imbalanced memory usage problem?

**Comment by myleott** [comment]
> Why not calculate the loss and gradient separately for each GPU, then all_reduce the gradients?

This is exactly what we do. Each worker independently computes gradients on a batch of data and we all-reduce them via PyTorch's DistributedDataParallel: https://pytorch.org/docs/master/nn.html#torch.nn.parallel.DistributedDataParallel
